File: backend/app/providers/openai_provider.py
from typing import Any, AsyncIterator, cast

from openai import AsyncOpenAI, AsyncStream
from openai.types.chat import (
    ChatCompletion,
    ChatCompletionChunk,
    ChatCompletionMessageParam,
)

from app.providers.base import ProviderChunk
from app.schemas.chat import ChatMessageSchema


def _to_openai_messages(
    messages: list[ChatMessageSchema],
) -> list[ChatCompletionMessageParam]:
    # The OpenAI SDK expects a union of typed message params; our internal
    # schema uses a simpler role/content shape.
    return cast(
        list[ChatCompletionMessageParam],
        [{"role": m.role, "content": m.content} for m in messages],
    )


def _coerce_message_content(content: Any) -> str:
    if isinstance(content, str):
        return content
    if content is None:
        return ""

    # Some SDK variants can return structured content parts.
    collected: list[str] = []
    for part in content:
        text = getattr(part, "text", None)
        if isinstance(text, str) and text:
            collected.append(text)
    return "".join(collected)


class OpenAIProvider:
    """LLMProvider adapter backed by the OpenAI Chat Completions API."""

    def __init__(self, api_key: str | None) -> None:
        self._client = AsyncOpenAI(api_key=api_key)

    async def stream_chat(
        self,
        messages: list[ChatMessageSchema],
        model: str,
        temperature: float = 0.7,
    ) -> AsyncIterator[ProviderChunk]:
        stream: AsyncStream[ChatCompletionChunk] = (
            await self._client.chat.completions.create(
                model=model,
                messages=_to_openai_messages(messages),
                temperature=temperature,
                stream=True,
            )
        )
        async for event in stream:
            if not event.choices:
                continue
            choice = event.choices[0]
            yield ProviderChunk(
                content=choice.delta.content or "",
                finish_reason=choice.finish_reason,
            )

    async def complete_chat(
        self,
        messages: list[ChatMessageSchema],
        model: str,
        temperature: float = 0.7,
    ) -> str:
        response: ChatCompletion = await self._client.chat.completions.create(
            model=model,
            messages=_to_openai_messages(messages),
            temperature=temperature,
            stream=False,
        )
        if not response.choices:
            return ""
        return _coerce_message_content(response.choices[0].message.content)

File: backend/nodejs/src/providers/openaiProvider.ts
import OpenAI from 'openai';

import type { ChatMessage, LLMProvider, ProviderChunk } from './base.js';

type OpenAIMessage = {
  role: ChatMessage['role'];
  content: string;
};

type OpenAIChoice = {
  message?: {
    content?: unknown;
  };
  delta?: {
    content?: string | null;
  };
  finish_reason?: string | null;
};

type OpenAICompletionResponse = {
  choices?: OpenAIChoice[];
};

type OpenAIStreamingResponse = AsyncIterable<OpenAICompletionResponse>;

type OpenAIClient = {
  chat: {
    completions: {
      create(
        request: Record<string, unknown>,
        signal?: AbortSignal,
      ): Promise<OpenAICompletionResponse | OpenAIStreamingResponse>;
    };
  };
};

function createOpenAIClient(apiKey: string): OpenAIClient {
  const client = new OpenAI({ apiKey });

  return {
    chat: {
      completions: {
        create(request, signal) {
          return client.chat.completions.create(
            request as unknown as Parameters<
              typeof client.chat.completions.create
            >[0],
            signal ? { signal } : undefined,
          ) as Promise<OpenAICompletionResponse | OpenAIStreamingResponse>;
        },
      },
    },
  };
}

function toOpenAIMessages(messages: ChatMessage[]): OpenAIMessage[] {
  return messages.map((message) => ({
    role: message.role,
    content: message.content,
  }));
}

function coerceMessageContent(content: unknown): string {
  if (typeof content === 'string') {
    return content;
  }

  if (!Array.isArray(content)) {
    return '';
  }

  return content
    .map((part) => {
      if (
        typeof part === 'object' &&
        part !== null &&
        'text' in part &&
        typeof part.text === 'string'
      ) {
        return part.text;
      }

      return '';
    })
    .join('');
}

export class OpenAIProvider implements LLMProvider {
  private readonly client: OpenAIClient;

  constructor(apiKey: string, client?: OpenAIClient) {
    this.client = client ?? createOpenAIClient(apiKey);
  }

  async completeChat(
    messages: ChatMessage[],
    model: string,
    temperature = 0.7,
  ): Promise<string> {
    const response = (await this.client.chat.completions.create({
      model,
      messages: toOpenAIMessages(messages),
      temperature,
      stream: false,
    })) as OpenAICompletionResponse;

    return coerceMessageContent(response.choices?.[0]?.message?.content);
  }

  async *streamChat(
    messages: ChatMessage[],
    model: string,
    temperature = 0.7,
    signal?: AbortSignal,
  ): AsyncIterable<ProviderChunk> {
    const stream = (await this.client.chat.completions.create(
      {
        model,
        messages: toOpenAIMessages(messages),
        temperature,
        stream: true,
      },
      signal,
    )) as OpenAIStreamingResponse;

    for await (const event of stream) {
      const choice = event.choices?.[0];

      if (!choice) {
        continue;
      }

      yield {
        content: choice.delta?.content ?? '',
        finishReason: choice.finish_reason ?? null,
      };
    }
  }
}
